lib/srv/alpnproxy/conn.go

```diff
@@ -87,3 +89,109 @@ func (conn readOnlyConn) RemoteAddr() net.Addr               { return &utils.Net
 func (conn readOnlyConn) SetDeadline(t time.Time) error      { return nil }
 func (conn readOnlyConn) SetReadDeadline(t time.Time) error  { return nil }
 func (conn readOnlyConn) SetWriteDeadline(t time.Time) error { return nil }
+
+// newPingConn returns a ping connection wrapping the provided net.Conn.
+func newPingConn(conn net.Conn) *pingConn {
+	return &pingConn{Conn: conn}
+}
+
+// pingConn wraps a net.Conn and add ping capabilities to it, including the
+// `WritePing` function and `Read` (which excludes ping packets).
+//
+// When using this connection, the packets written will contain an initial data:
+// the packet size. When reading, this information is taken into account, but it
+// is not returned to the caller.
+//
+// Ping messages have a packet size of zero and are produced only when
+// `WritePing` is called. On `Read`, any Ping packet is discarded.
+type pingConn struct {
+	net.Conn
+
+	muRead  sync.Mutex
+	muWrite sync.Mutex
+
+	// bytesRead number of bytes already read from the current packet.
+	bytesRead int
+	// currentSize size of bytes of the current packet.
+	currentSize int32
+}
+
+func (c *pingConn) Read(p []byte) (int, error) {
+	c.muRead.Lock()
+	defer c.muRead.Unlock()
```

Comment:
It is for the case we have concurrent `Read` calls, I know it is not that common, but it seems safer to handle those scenarios. Without the lock, we would have data race for `bytesRead` and `currentSize`. This could cause the `Read` calls to try to read ping packets where they would need to read for data content and vice-versa.

There is a `ConcurrentRead` test case to cover those scenarios.